Make CreateProject work in native mode

**Describe the bug**
I've tried to setup Code Quarkus to compile to native with master.
quarkusio/code.quarkus.io#194

When generating an application, it fails with an exception:
```
org.jboss.resteasy.spi.UnhandledException: java.nio.file.ProviderNotFoundException: Provider not found
        at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
        at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
        at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:209)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:496)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:252)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:153)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:363)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:156)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:238)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:69)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:104)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatchRequestContext(VertxRequestHandler.java:83)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.lambda$handle$0(VertxRequestHandler.java:70)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:316)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:460)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
Caused by: java.nio.file.ProviderNotFoundException: Provider not found
        at java.nio.file.FileSystems.newFileSystem(FileSystems.java:407)
        at io.quarkus.platform.descriptor.loader.json.impl.QuarkusJsonPlatformDescriptor.loadResource(QuarkusJsonPlatformDescriptor.java:130)
        at io.quarkus.maven.utilities.MojoUtils.getProperties(MojoUtils.java:63)
        at io.quarkus.maven.utilities.MojoUtils.getAllProperties(MojoUtils.java:81)
        at io.quarkus.cli.commands.CreateProject.doCreateProject(CreateProject.java:107)
        at io.quarkus.code.services.QuarkusProjectCreator.create(QuarkusProjectCreator.kt:39)
        at io.quarkus.code.CodeQuarkusResource.download(CodeQuarkusResource.kt:76)
        at io.quarkus.code.CodeQuarkusResource_Subclass.download$$superaccessor2(CodeQuarkusResource_Subclass.zig:22)
        at io.quarkus.code.CodeQuarkusResource_Subclass$$function$$2.apply(CodeQuarkusResource_Subclass$$function$$2.zig:51)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
        at io.quarkus.hibernate.validator.runtime.interceptor.AbstractMethodValidationInterceptor.validateMethodInvocation(AbstractMethodValidationInterceptor.java:67)
        at io.quarkus.hibernate.validator.runtime.jaxrs.JaxrsEndPointValidationInterceptor.validateMethodInvocation(JaxrsEndPointValidationInterceptor.java:29)
        at io.quarkus.hibernate.validator.runtime.jaxrs.JaxrsEndPointValidationInterceptor_Bean.intercept(JaxrsEndPointValidationInterceptor_Bean.zig:106)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
        at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
        at io.quarkus.code.CodeQuarkusResource_Subclass.download(CodeQuarkusResource_Subclass.zig:218)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:151)
        at org.jboss.resteasy.core.MethodInjectorImpl.lambda$invoke$3(MethodInjectorImpl.java:122)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
        at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:614)
        at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1983)
        at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:110)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:122)
        at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:594)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:468)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:421)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:363)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:423)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:391)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invoke$1(ResourceMethodInvoker.java:365)
        at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:981)
        at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2124)
        at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:110)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:365)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:477)
        ... 16 more
```

To reproduce:
1. build in native
2. run
3. hit http://0.0.0.0:8080/api/download?g=org.acme&a=code-with-quarkus&v=1.0.0-SNAPSHOT&c=org.acme.ExampleResource&e=io.quarkus%3Aquarkus-smallrye-openapi&e=io.quarkus%3Aquarkus-resteasy

--- a/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/ClassPathResourceLoader.java
+++ b/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/ClassPathResourceLoader.java
@@ -1,0 +1,22 @@
+package io.quarkus.platform.descriptor.loader.json.impl;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+public class ClassPathResourceLoader implements ResourceLoader {
+
+    private final ClassLoader cl;
+
+    public ClassPathResourceLoader(ClassLoader cl) {
+        this.cl = cl;
+    }
+
+    @Override
+    public InputStream getResourceAsStream(String name) throws IOException {
+        final InputStream stream = cl.getResourceAsStream(name);
+        if (stream == null) {
+            throw new IOException("Failed to locate " + name + " on the classpath");
+        }
+        return stream;
+    }
+}

--- a/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/DirectoryResourceLoader.java
+++ b/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/DirectoryResourceLoader.java
@@ -1,0 +1,30 @@
+package io.quarkus.platform.descriptor.loader.json.impl;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class DirectoryResourceLoader implements ResourceLoader {
+
+    private final Path dir;
+
+    public DirectoryResourceLoader(Path dir) {
+        if (!Files.isDirectory(dir)) {
+            throw new IllegalStateException("Failed to locate directory " + dir);
+        }
+        this.dir = dir;
+    }
+
+    @Override
+    public InputStream getResourceAsStream(String name) throws IOException {
+        Path resource = dir.resolve(name);
+        if (!Files.exists(resource)) {
+            throw new IOException("Failed to locate " + resource);
+        }
+        if (Files.isDirectory(resource)) {
+            throw new IOException("Can't open a stream for path pointing to directory " + resource);
+        }
+        return Files.newInputStream(resource);
+    }
+}

--- a/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/QuarkusJsonPlatformDescriptor.java
+++ b/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/QuarkusJsonPlatformDescriptor.java
@@ -3,17 +3,13 @@ package io.quarkus.platform.descriptor.loader.json.impl;
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStream;
-import java.nio.file.FileSystem;
-import java.nio.file.FileSystems;
-import java.nio.file.Files;
-import java.nio.file.Path;
+import java.io.InputStreamReader;
 import java.util.Collections;
 import java.util.List;
 import java.util.stream.Collectors;
 
 import org.apache.maven.model.Dependency;
 
-import io.quarkus.bootstrap.util.ZipUtils;
 import io.quarkus.dependencies.Category;
 import io.quarkus.dependencies.Extension;
 import io.quarkus.platform.descriptor.QuarkusPlatformDescriptor;
@@ -31,7 +27,7 @@ public class QuarkusJsonPlatformDescriptor implements QuarkusPlatformDescriptor 
     private List<Extension> extensions = Collections.emptyList();
     private List<Dependency> managedDeps = Collections.emptyList();
     private List<Category> categories = Collections.emptyList();
-    private Path templatesJar;
+    private ResourceLoader resourceLoader;
     private MessageWriter log;
 
     public QuarkusJsonPlatformDescriptor() {
@@ -51,8 +47,8 @@ public class QuarkusJsonPlatformDescriptor implements QuarkusPlatformDescriptor 
         this.managedDeps = managedDeps;
     }
 
-    void setTemplatesJar(Path templatesJar) {
-        this.templatesJar = templatesJar;
+    void setResourceLoader(ResourceLoader resourceLoader) {
+        this.resourceLoader = resourceLoader;
     }
 
     void setMessageWriter(MessageWriter log) {
@@ -100,41 +96,26 @@ public class QuarkusJsonPlatformDescriptor implements QuarkusPlatformDescriptor 
     @Override
     public String getTemplate(String name) {
         getLog().debug("Loading Quarkus project template %s", name);
-        if (templatesJar == null) {
-            return null;
+        if (resourceLoader == null) {
+            throw new IllegalStateException("Resource loader has not been provided");
         }
-        try {
-            if (Files.isDirectory(templatesJar)) {
-                return readTemplate(name, templatesJar.resolve(name));
-            }
-            try (FileSystem fs = ZipUtils.newFileSystem(templatesJar)) {
-                return readTemplate(name, fs.getPath(name));
+        try (InputStream is = resourceLoader.getResourceAsStream(name)) {
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {
+                return reader.lines().collect(Collectors.joining("\n"));
             }
         } catch (IOException e) {
             throw new RuntimeException("Failed to resolve template " + name, e);
         }
     }
 
-    private String readTemplate(String name, final Path p) throws IOException {
-        if (!Files.exists(p)) {
-            throw new RuntimeException("Failed to locate template " + name + " in " + templatesJar);
-        }
-        try (BufferedReader reader = Files.newBufferedReader(p)) {
-            return reader.lines().collect(Collectors.joining("\n"));
-        }
-    }
-
     @Override
     public <T> T loadResource(String name, ResourceInputStreamConsumer<T> consumer) throws IOException {
         getLog().debug("Loading Quarkus platform resource %s", name);
-        try (FileSystem fs = FileSystems.newFileSystem(templatesJar, null)) {
-            final Path p = fs.getPath(name);
-            if (!Files.exists(p)) {
-                throw new IOException("Failed to locate resource " + name);
-            }
-            try (InputStream is = Files.newInputStream(p)) {
-                return consumer.handle(is);
-            }
+        if (resourceLoader == null) {
+            throw new IllegalStateException("Resource loader has not been provided");
+        }
+        try (InputStream is = resourceLoader.getResourceAsStream(name)) {
+            return consumer.handle(is);
         }
     }
 

--- a/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/QuarkusJsonPlatformDescriptorLoaderImpl.java
+++ b/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/QuarkusJsonPlatformDescriptorLoaderImpl.java
@@ -3,6 +3,7 @@ package io.quarkus.platform.descriptor.loader.json.impl;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
 
@@ -75,11 +76,22 @@ public class QuarkusJsonPlatformDescriptorLoaderImpl
             throw new RuntimeException("Failed to determine the Quarkus version for the platform " + platformBom);
         }
 
+        final Path classOrigin;
         try {
-            platform.setTemplatesJar(MojoUtils.getClassOrigin(getClass()));
+            classOrigin = MojoUtils.getClassOrigin(getClass());
         } catch (Exception e) {
             throw new IllegalStateException("Failed to determine the origin of " + getClass().getName(), e);
         }
+
+        final ResourceLoader resourceLoader;
+        if (Files.isDirectory(classOrigin)) {
+            resourceLoader = new DirectoryResourceLoader(classOrigin);
+        } else {
+            // this means the class belongs to a JAR which is on the classpath
+            resourceLoader = new ClassPathResourceLoader(getClass().getClassLoader());
+        }
+        platform.setResourceLoader(resourceLoader);
+
         platform.setQuarkusVersion(quarkusVersion);
         platform.setMessageWriter(context.getMessageWriter());
 

--- a/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/ResourceLoader.java
+++ b/devtools/platform-descriptor-json/src/main/java/io/quarkus/platform/descriptor/loader/json/impl/ResourceLoader.java
@@ -1,0 +1,9 @@
+package io.quarkus.platform.descriptor.loader.json.impl;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+public interface ResourceLoader {
+
+    InputStream getResourceAsStream(String name) throws IOException;
+}

